### PR TITLE
rustc-book: Document `-C target-feature=+crt-static`

### PR DESCRIPTION
**Pull request by petrochenkov**
A preliminary for https://github.com/rust-lang/rust/pull/70740.

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @alexcrichton

File: src/doc/rustc/src/codegen-options/index.md [hunk 1]
```diff
@@ -367,12 +401,35 @@ machine. Each target has a default base CPU.
...
+compilation.  
+Code configured with `cfg(target_feature = "x")` is kept if the feature `x` is enabled,
+and removed otherwise.
+
+If the enabled feature is not supported for the current target, the behavior is unspecified.
+<!-- There are multiple bugs here
+  - Unknown CPU features go straight to LLVM and appear as LLVM warnings.
+  - At configure time `crt-static` predicate is set even if the target doesn't support `crt-static`.
+  - At link time `crt-static` is sometimes ignored
+    if the target doesn't support it and sometimes not.
```
Comment (petrochenkov) [inline src/doc/rustc/src/codegen-options/index.md:421]: Also, I strongly suspect that https://github.com/rust-lang/rust/pull/69519 was a hack rather than a proper solution, but I need some time to prepare a musl setup and verify things. (Proc-macros shouldn't be any different to cdylibs and dylibs with regards to linking behavior.)

**Comment by steveklabnik** [comment]
This includes more than just documentation changes, but also implementation changes ,was that intended?

**Comment by petrochenkov** [comment]
@steveklabnik 
>Based on https://github.com/rust-lang/rust/pull/71490, only the last commit is new.

File: src/doc/rustc/src/codegen-options/index.md [hunk 2]
```diff
@@ -381,11 +438,53 @@ See also the [`target_feature`
...
+
+- PE-based executables (Windows).  
+The produced executable will contain code from all the user-level libraries, including
+[C Run-Time Libraries (CRT)](https://docs.microsoft.com/en-us/cpp/c-runtime-library/crt-library-features),
+, but will still link dynamically to system libraries like `kernel32.dll`.
+
+- PE-based shared libraries (Windows).  
+The produced shared library will contain code from all the user-level libraries, including
+[C Run-Time Libraries (CRT)](https://docs.microsoft.com/en-us/cpp/c-runtime-library/crt-library-features),
+, but will still link dynamically to system libraries like `kernel32.dll`.
```
Comment (ehuss) [inline src/doc/rustc/src/codegen-options/index.md:477]:
```suggestion
but will still link dynamically to system libraries like `kernel32.dll`.
```

File: src/doc/rustc/src/codegen-options/index.md [hunk 3]
```diff
@@ -367,12 +401,35 @@ machine. Each target has a default base CPU.
...
+  - At link time `crt-static` is sometimes ignored
+    if the target doesn't support it and sometimes not.
+-->
+
+To see the valid features and an example of use, run `rustc --print target-features`.
+
+#### CPU instruction set extensions
+
+Target features that allow enabling or disabling architecture-dependent CPU features,
+like `+neon` or `-sse`.
```
Comment (ehuss) [inline src/doc/rustc/src/codegen-options/index.md:429]: This sentence seems a little strange without a verb.  Maybe something like this?

> Target features for CPU instruction set extensions can enable or disable architecture-dependent CPU features, like `+neon` or `-sse`.

**Review by ehuss — COMMENTED** [review]
Thanks for updating the docs, they are a big improvement!

File: src/doc/rustc/src/codegen-options/index.md (hunk 1 above)
Comment (mati865) [reply src/doc/rustc/src/codegen-options/index.md:422]: > Proc-macros shouldn't be any different to cdylibs and dylibs with regards to linking behavior

That was the problem 😉

TL;DR It's a hack.

Musl target is a "weirdo". Because of the legacy decisions it defaults to static linking so it cannot produce dylibs, cdylibs and proc-macros. Now lack of proc-macro support doesn't matter when cross-compiling from the host which supports them to musl.
It becomes difficult when using musl as the host toolchain. Since it defaults to static linking proc-macros don't work. To build them one had to use `target-feature=-crt-static` and lose ability to statically link binaries.

**Comment by alexcrichton** [comment]
I haven't looked too too closely at this, but my main recommendation would be to not try to exhaustively document the state of things without a disclaimer. There's basically always one reason or another that `+/-crt-static` is still buggy, so I think it'd be worthwhile to place a disclaimer in the docs saying that things can be tweaked over time and if someone is overly surprised it's recommended they file an issue.

File: src/doc/rustc/src/codegen-options/index.md (hunk 1 above)
Comment (petrochenkov) [reply src/doc/rustc/src/codegen-options/index.md:422]: @mati865 
First, musl toolchain actually supports statically linking self-contained `.so`s, despite the flag in `rustc` target spec saying otherwise, I tried it (at least with "hello world"-like examples).

Second, that doesn't mean that we *want* to use it for proc macros (as well as cdylibs and dylibs) at the same time as we are building a statically linked executable.

One common `RUSTFLAGS="-Ctarget-feature=+crt-static"` (or a target spec default) in this case brings collateral damage on the crates which it is not supposed to apply to.

I see multiple possible solutions here some of which could be more principled than what https://github.com/rust-lang/rust/pull/69519 did.
- Automatically enable `-Ctarget-feature=-crt-static` for proc macro crates or all libraries in Cargo, modifying RUSTFLAGS. Question: how to opt-out?
- Introduce a new options `-Ctarget-feature=+crt-static-dylib` controlling static linking of libraries instead of `-Ctarget-feature=-crt-static`. It would almost never be used on Linux (not sure about windows-msvc and wasm).
- Keep the existing meaning of `crt-static`, but introduce a new option `-C disable-crt-static-for-dylibs` or something. Cargo would then use it for proc macro crates or all libraries. Question: how to opt-out?
- Ignore `+crt-static` for dylibs if the target doesn't support it. This is fragile, musl actually supports it despite the current value of the flag of the musl target spec. If the flag is enabled, proc macros will break.
- Ignore `+crt-static` for dylibs always. They are almost never used on Linux (not sure about windows-msvc and wasm). `+crt-static-dylib` looks strictly better.

Comment (petrochenkov) [reply src/doc/rustc/src/codegen-options/index.md:422]: - Have two `crt-static` defaults in target specs - one for executables and one for libraries. Solves one half of the problem, but explicit `+crt-static` in RUSTFLAGS will still cause collateral damage.

File: src/doc/rustc/src/codegen-options/index.md [hunk 4]
```diff
@@ -416,11 +439,53 @@ See also the [`target_feature`
...
+
+- ELF-based shared libraries.  
+If the target supports statically linked shared libraries,
+then the linker will be instructed (`-shared -static`) to produce one.
+The shared library will contain code from all the used libraries, including `libc`, inside it.
+This option is not widely supported or used.
+
+- PE-based executables (Windows).  
+The produced executable will contain code from all the user-level libraries, including
+[C Run-Time Libraries (CRT)](https://docs.microsoft.com/en-us/cpp/c-runtime-library/crt-library-features),
```
Comment (retep998) [inline src/doc/rustc/src/codegen-options/index.md:472]: What user-level libraries will it statically link to other than the CRT? For `windows-msvc` `crt-static` *only* controls whether the CRT (consisting of the VC Runtime and the UCRT) is statically linked and nothing else is affected. Also it only applies to `windows-msvc` while `windows-gnu` does not support `crt-static`.

Comment (petrochenkov) [reply src/doc/rustc/src/codegen-options/index.md:472]: @retep998 
>What user-level libraries will it statically link to other than the CRT?

Rust crates.
And it should do that for native dependencies as well (https://github.com/rust-lang/rust/issues/71647), but that doesn't make any difference on Windows since there's no `libmylib.a`/`libmylib.so` separation.

If after that anything is still linked dynamically, then we get into the `*WARNING!*` clause below. It's not as dangerous in the Windows case though, since the goal there is not to produce a fully statically linked executable, like in the ELF case.

>Also it only applies to `windows-msvc` while `windows-gnu` does not support `crt-static`.

Good catch.
This paragraph needs the same "If the target supports..." condition as the ELF paragraphs above.

File: src/doc/rustc/src/codegen-options/index.md (hunk 1 above)
Comment (petrochenkov) [reply src/doc/rustc/src/codegen-options/index.md:422]: I'll turn this into a proper issue.

**Comment by petrochenkov** [comment]
TODO: Mention that `+crt-static` overpowers `-C prefer-dynamic`.

File: src/doc/rustc/src/codegen-options/index.md (hunk 1 above)
Comment (petrochenkov) [reply src/doc/rustc/src/codegen-options/index.md:422]: https://github.com/rust-lang/rust/issues/71651

**Comment by petrochenkov** [comment]
This is kind of blocked on https://github.com/rust-lang/rust/issues/71651 now.

File: src/doc/rustc/src/codegen-options/index.md (hunk 4 above)
Comment (retep998) [reply src/doc/rustc/src/codegen-options/index.md:472]: To me `crt-static` on Windows always had the impression that the only thing it controlled (and ever intended to control) was whether it linked to the static or dynamic CRT. I was not aware that it had any effect on whether a rust crate was linked statically or dynamically, as `-Cprefer-dynamic` already covers that.

**Comment by petrochenkov** [comment]
TODO: Figure out what gcc/clang do on Linux if libc is linked statically, but something else is linked dynamically.

**Comment by smaeul** [comment]
> TODO: Figure out what gcc/clang/ld/lld do on Linux if libc is linked statically, but something else is linked dynamically.

We send both `-static` and `-Bdynamic` to the linker, causing it to produce a nonfunctional dynamic ELF.

From https://github.com/rust-lang/rust/pull/55566#issuecomment-436095958 quoting https://github.com/rust-lang/rust/pull/40113#issuecomment-323193341:

> 4) The current problem is that, when `base.dynamic_linking = true`, `base.crt_static_default = true` *is not enough to enforce that binaries are statically linked by default*.
>    * It does not enforce that all linked libraries must also be statically linked, so when `gcc` sees dynamic libraries, it creates a dynamic executable (i.e. an ELF with `DT_NEEDED` entries).
>    * But because *you told it* you were creating a static executable, `gcc` not pass `-dynamic-linker` to `ld` to add the correct `PT_INTERP` header to the ELF, and therefore `ld` uses its hardcoded default for the architecture, e.g. `/lib/ld64.so.1` on x86_64.

**Comment by petrochenkov** [comment]
Bibliography :slightly_smiling_face::
https://github.com/rust-lang/rust/issues/11937 - Link using the linker directly
https://github.com/rust-lang/rust/issues/31322 - How to compile rust in a musl libc based linux system
https://github.com/rust-lang/rust/issues/34987 - MUSL dynamic linking support
https://github.com/rust-lang/libc/issues/290 - Can't statically link MSVC runtime
https://github.com/rust-lang/rfcs/pull/1684 - Choosing which CRT to link to
https://github.com/rust-lang/rfcs/pull/1721 - Enable customizing the linkage of a platform's C runtime
https://github.com/rust-lang/rust/issues/36710 - link: failed to static link to c++ library when global variable is used
https://github.com/rust-lang/rust/issues/37406 - Tracking issue for customizing CRT linkage
https://github.com/rust-lang/rust/pull/37545 - rustc: Implement #[link(cfg(..))] and crt-static
https://github.com/rust-lang/rust/pull/38075 - x86_64-musl: allow building dylibs with -crt-static
https://github.com/rust-lang/rust/issues/39998 - A new compiler flag: "link everything statically or die, dammit!"
https://github.com/rust-lang/rust/pull/40113 - Support dynamically-linked and/or native musl targets
https://github.com/rust-lang/rust/issues/40174 - Unable to find crate `proc_macro` on musl target
https://internals.rust-lang.org/t/refining-cross-platform-crt-static-semantics/5085 - Refining cross-platform crt-static semantics
https://github.com/rust-lang/rust/pull/50105 - Use the correct crt*.o files when linking musl targets
https://github.com/rust-lang/rust/pull/50709 - Revert #50105 until regression is fixed
https://github.com/rust-lang/rust/pull/51063 - Fix building rustc on and for musl hosts
https://github.com/rust-lang/rust/issues/53968 - static-pie for musl target (remove crt1.o)
https://github.com/rust-lang/rust/issues/54243 - x86_64-unknown-linux-musl target and main-returning-Result produces a dynamically-linked executable instead of a statically linked one
https://github.com/rust-lang/rust/pull/55566 - Require static native libraries when linking static executables
https://github.com/rust-lang/rust/pull/58575 - Musl host toolchain
https://github.com/rust-lang/rust/issues/59302 - Tracking issue for musl host toolchain
https://github.com/rust-lang/rust/issues/64547 - mips-linux-musleab is broken again because it uses the wrong crt for static pie
https://github.com/rust-lang/rust/issues/65447 - Statically link against glibc
https://github.com/rust-lang/rust/issues/65760 - musl staticlib builds include all of the musl symbols
https://github.com/rust-lang/rust/pull/69519 - Don't use static crt by default when build proc-macro
https://github.com/rust-lang/rust/issues/70693 - RFE: x86_64-unknown-linux-musl `static-pie` target for ASLR
https://github.com/rust-lang/rust/pull/71586 - rustc-book: Document `-C target-feature=+crt-static`
https://github.com/rust-lang/rust/issues/71647 - `rustc` should prefer statically linking native dependencies with unspecified kind if `crt-static` is enabled
https://github.com/rust-lang/rust/issues/71651 - Enabling `+crt-static` in a blanket way breaks dynamic libraries including proc macros
https://github.com/rust-lang/rust/pull/71804 - linker: Support `-static-pie` and `-static -shared`
https://github.com/rust-lang/libc/pull/1327 - Don't bundle static libc
https://github.com/rust-lang/cargo/issues/7563 - cannot produce proc-macro when `--target` is set on musl host toolchain
https://gcc.gnu.org/legacy-ml/gcc/2015-06/msg00008.html - Static PIE support in GCC
https://www.openwall.com/lists/musl/2012/12/08/2 - static linking and dlopen
https://www.openwall.com/lists/musl/2016/07/14/1 - Regarding whole-archive linking libc.a into a shared lib

**Comment by petrochenkov** [comment]
I'll reopen this once the related issues are addressed.
